nixos/modules/services/web-apps/opencloud.nix

```diff
@@ -0,0 +1,212 @@
+        type = types.attrsOf types.str;
+        default = { };
+        description = ''
+          Extra config options.
+
+          Note that all the attributes here will be copied to /nix/store/ and will be world readable. Options like *_PASSWORD or *_SECRET should be part of     [environmentFile](#opt-services.opencloud.environmentFile) instead, and are only provided here for illustrative purpose.
+
+          Configuration here will override those from [environmentFile](#opt-services.opencloud.environmentFile) and will have highest precedence, at the cost of security. Do NOT put security sensitive stuff here.
+        '';
+        example = {
+          OC_INSECURE = "false";
+          OC_LOG_LEVEL = "error";
+          OC_JWT_SECRET = "super_secret";
+          OC_TRANSFER_SECRET = "foo";
+          OC_MACHINE_AUTH_API_KEY = "foo";
+          OC_SYSTEM_USER_ID = "123";
+          OC_MOUNT_ID = "123";
+          OC_STORAGE_USERS_MOUNT_ID = "123";
+          GATEWAY_STORAGE_USERS_MOUNT_ID = "123";
+          CS3_ALLOW_INSECURE = "true";
+          OC_INSECURE_BACKENDS = "true";
+          TLS_INSECURE = "true";
+          TLS_SKIP_VERIFY_CLIENT_CERT = "true";
+          WEBDAV_ALLOW_INSECURE = "true";
+          IDP_TLS = "false";
+          GRAPH_APPLICATION_ID = "1234";
+          IDM_IDPSVC_PASSWORD = "password";
+          IDM_REVASVC_PASSWORD = "password";
+          IDM_SVC_PASSWORD = "password";
+          IDP_ISS = "https://localhost:9200";
+          OC_LDAP_BIND_PASSWORD = "password";
+          OC_SERVICE_ACCOUNT_ID = "foo";
+          OC_SERVICE_ACCOUNT_SECRET = "foo";
+          OC_SYSTEM_USER_API_KEY = "foo";
+          STORAGE_USERS_MOUNT_ID = "123";
+        };
+      };
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    users.users.${defaultUser} = lib.mkIf (cfg.user == defaultUser) {
+      group = cfg.group;
+      home = cfg.stateDir;
+      isSystemUser = true;
+      createHome = true;
+      description = "OpenCloud daemon user";
+    };
+
+    users.groups = lib.mkIf (cfg.group == defaultGroup) { ${defaultGroup} = { }; };
+
+    systemd = {
+      services.opencloud = {
+        description = "OpenCloud - a secure and private way to store, access, and share your files";
+        wantedBy = [ "multi-user.target" ];
+        environment = {
+          PROXY_HTTP_ADDR = "${cfg.address}:${toString cfg.port}";
+          OC_URL = cfg.url;
+          OC_CONFIG_DIR = if (cfg.configDir == null) then "${cfg.stateDir}/config" else cfg.configDir;
+          OC_BASE_DATA_PATH = cfg.stateDir;
+          WEB_ASSET_CORE_PATH = "${cfg.webPackage}";
+          IDP_ASSET_PATH = "${cfg.idpWebPackage}/assets";
+        } // cfg.environment;
+        serviceConfig = {
+          Type = "simple";
+          ExecStart = "${lib.getExe cfg.package} server";
+          WorkingDirectory = cfg.stateDir;
+          User = cfg.user;
+          Group = cfg.group;
+          Restart = "always";
+          DynamicUser = true;
```

Comment:
Dynamic user should not store large amounts of files as they are chownes on every start